Extend `UnknownFlags` to `UnknownFlagsHandling` and introduce `PassUnknownFlagToArgs` (#337)

ikedam opened the pull request:
Close #337

Alternative to #338.

`FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling=PassUnknownFlagToArgs` behaves like this:

* Input:
    ```
    -xayb -c -z --known-flag known-flag-value --unknown-flag arg0 arg1 arg2
    ```
    * -x, -y, -z and  `--unknown-flag` are unknown
* Args() results:
    ```
    -xy -z --unknown-flag arg0 arg1 arg2
    ```


This request deprecates `FlagSet.ParseErrorsAllowlist.UnknownFlags` and extends it to `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling`.

* `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling = ErrorOnUnknownFlag` (Default)
    * Same to `FlagSet.ParseErrorsAllowlist.UnknownFlags = false`
    * Treats unknown flags error.
* `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling = IgnoreUnknownFlag`
    * Same to `FlagSet.ParseErrorsAllowlist.UnknownFlags = true`
    * Ignore unknown flags.
* `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling=PassUnknownFlagToArgs`:
    * Pass unknown flags to `Args()` (described above)

Backward compatibilities:

* `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling` is not set (`ErrorOnUnknownFlag`) and `FlagSet.ParseErrorsAllowlist.UnknownFlags = true`:
    * Same to `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling = ErrorOnUnknownFlag`
    * Ignore unknown flags.
* Also same for `ParseErrorsWhitelist`

ikedam committed:
Extend `UnknownFlags` to `UnknownFlagsHandling` and introduce `UnknownFlagsHandlingPassUnknownToArgs` (#337)

* Conflicted with #446

ikedam commented on flag_test.go:
```diff
@@ -523,6 +523,111 @@ func testParseWithUnknownFlags(f *FlagSet, t *testing.T) {
+	if f.Parsed() {
+		t.Fatal("f.Parse() = true before Parse")
+	}
+	f.ParseErrorsAllowlist.UnknownFlagsHandling = UnknownFlagsHandlingPassUnknownToArgs
```
This test code is same to https://github.com/spf13/pflag/pull/338/files#diff-7a7385a10188cdaa9ac18effc375607e65672f55084b8049b6d3f44a6ab70c51 except this line.

tomasaschan commented on flag.go:
```diff
@@ -137,10 +137,42 @@ const (
+	// Combined shorthand flags mixed with known ones and unknown ones results
+	// combined flags only with unknown ones.
+	// E.g. -fghi results -gh if only `f` and `i` are known.
+	UnknownFlagsHandlingPassUnknownToArgs
```
Maybe `ErrorOnUnknownFlag`? Along with `IgnoreUnknownFlag` and `PassUnknownFlagToArgs`. I don't think they can be made shorter than that without losing meaning, but I don't think we need to be overly cautious about naming conflicts here.

If you really want to guard against it, another variant is to namespace these somehow - either with a _struct value_ where the enum values are fields on the struct, or by just moving them to an appropriately named subpackage.

ikedam replied:
Renamed in https://github.com/spf13/pflag/pull/440/changes/c0bdd631c7115eaabfce7d6c38f09cb42da8909e

tomasaschan commented on flag.go:
```diff
@@ -137,10 +137,42 @@ const (
+	if a.UnknownFlags {
+		return UnknownFlagsHandlingIgnoreUnknown
+	}
+	return UnknownFlagsHandlingErrorOnUnknown
```
Nice!

tomasaschan reviewed (commented):
I like this version a lot better than #338.

(The passing-error-as-control-flow thing smells a little, but the `parseLongArg` method was a mess before this change, so that's not your fault... Something to look out for in 2.0, I guess.)

Needs a rebase (I haven't looked into why) and probably shouldn't be merged until we've decided we don't have anything else we want to get into 1.0 before cutting 1.1.

ikedam committed:
Handle both ParseErrorsAllowlist and ParseErrorsWhitelist #337

ikedam commented:
First, rebased and resolved the conflict.
Conflicted with #446 .

Needs to support both `ParseErrorsAllowlist` and `ParseErrorsWhitelist`, and resolved as https://github.com/spf13/pflag/pull/440/changes/d1d7aeebabe5cb8f98a083bc82685272c6083c1b .
The conflict was left unresolved in the previous commit and fixed in this one so reviewers can see exactly how it was resolved.

ikedam committed:
Rename symbols to shorter ones #337

* `UnknownFlagsHandlingErrorOnUnknown` to `ErrorOnUnknownFlag` and so on

tomasaschan commented on flag.go:
```diff
@@ -999,22 +1058,25 @@ func (f *FlagSet) parseLongArg(s string, args []string, fn parseFunc) (a []strin
 				return a, nil
 			}
 
 			return stripUnknownFlagValue(a), nil
+		case unknownFlagsHandling == PassUnknownFlagToArgs:
+			return a, &unknownFlagError{
```
I don't understand why the previous case calls `stripUnknownFlagValue` and this is one does not.

ikedam replied:
We can consider the case of `--unknown somevalue`.
`--unknown` is an unknown flag, and we cannot reliably tell whether `somevalue` is a flag value or a positional argument.

With `PassUnknownFlagToArgs`, we do not need to classify `somevalue`; we simply leave the argument stream intact and let the caller receive everything as positional args. So we don't need to call `stripUnknownFlagValue()`.

With `IgnoreUnknownFlag`, the existing behavior treats the next token as a flag value (when it does not look like a flag) and calls `stripUnknownFlagValue()` to strip it. This behavior is inherited from the existing `IgnoreUnknownFlag` implementation.

tomasaschan commented on flag.go:
```diff
@@ -1077,6 +1139,20 @@ func (f *FlagSet) parseSingleShortArg(shorthands string, args []string, fn parse
 			return
+		case unknownFlagsHandling == PassUnknownFlagToArgs:
+			// '-f=arg': pass all the argument
+			if len(shorthands) > 2 && shorthands[1] == '=' {
+				outShorts = ""
+				err = &unknownFlagError{
+					UnknownFlags: shorthands,
+				}
+				return
+			}
+			// '-fgh': pass only the first switch
+			err = &unknownFlagError{
+				UnknownFlags: shorthands[0:1],
+			}
+			return
```
What about `-fg=g` were both f and g are shorthands, but only g accept a value? Or `-fgh` (same)?

ikedam replied:
Same here, differs in the next call.

For an unknown flag, `parseShortArg` only need to choose between:

* strip only the first letter, or
* strip all the argument

Both `-fg=g` and `-fgh` are the former case: only strips the first letter (`f`) as an unknown flag, and pass the rest (`g=g` or `gh`).
The rest will be handled in the next `parseSingleShortArg()` call.

In the next `parseSingleShortArg()`, `g=g` results the latter case, and `gh` results the former case.

tomasaschan reviewed (approved):
I'm sorry this took so long to get to reviewing. I have a couple of questions, but given this behavior is opt-in I think it would be OK to merge this to make it easier to test in real applications, and roll forward when more edge cases pop up.

Let me know if you prefer a merge now or to address those questions first.
